TE-17.1 : vrf_policy_driven_te_test.go

cprabha opened the pull request:
Hi,

Initial commit for TE-17.1 test.
Resolves: https://github.com/openconfig/featureprofiles/issues/2222

Needs to merge https://github.com/openconfig/featureprofiles/pull/2539 for ip address update.

Needs https://github.com/openconfig/featureprofiles/pull/2591 for 8 port topology.

Thanks,
Prabha

sezhang2 commented on feature/gribi/otg_tests/vrf_policy_driven_te/vrf_policy_driven_te_test.go:
```diff
@@ -0,0 +1,1946 @@
+}
+
+// testGribiDecapMatchSrcProtoNoMatchDSCP is to validate subtest test1.
+// Test-1 match on source and protocol no match on DSCP; flow VRF_DECAP hit -> DEFAULT
```
Can you write the test cases in the same order as listed in README.md file. For example, Test 1 should be the 1st one.

cprabha replied:
I have updated testcase order as per README. please check.

sezhang2 reviewed (commented):
Can you move the test code to expremental folder as tests could not pass
 due to Ixia issue in function   _testTunnelTrafficDecapEncap_?

- https://github.com/open-traffic-generator/fp-testbed-juniper/issues/42

Please also fix the static analysis errors and address the comment.

arvbaska1 commented on feature/gribi/otg_tests/vrf_policy_driven_te/vrf_policy_driven_te_test.go:
```diff
@@ -0,0 +1,1946 @@
+		fluent.NextHopGroupEntry().WithNetworkInstance(deviations.DefaultNetworkInstance(dut)).
+			WithID(4).AddNextHop(3, 2).WithBackupNHG(19),
+		fluent.IPv4Entry().WithNetworkInstance(niTeVrf111).
+			WithPrefix(gribiIPv4EntryVRF1112+"/"+maskLen32).WithNextHopGroup(4),
```
Similarly please add WithNextHopGroupNetworkInstance(deviations.DefaultNetworkInstance(dut) here as well

cprabha replied:
automated as per README

arvbaska1 commented on feature/gribi/otg_tests/vrf_policy_driven_te/vrf_policy_driven_te_test.go:
```diff
@@ -0,0 +1,1946 @@
+		fluent.NextHopGroupEntry().WithNetworkInstance(deviations.DefaultNetworkInstance(dut)).
+			WithID(1).AddNextHop(1, 1).AddNextHop(2, 3).WithBackupNHG(18),
+		fluent.IPv4Entry().WithNetworkInstance(niTeVrf111).
+			WithPrefix(gribiIPv4EntryVRF1111+"/"+maskLen32).WithNextHopGroup(1),
```
Similarly please add WithNextHopGroupNetworkInstance(deviations.DefaultNetworkInstance(dut) here as well

cprabha replied:
same as above

arvbaska1 commented on feature/gribi/otg_tests/vrf_policy_driven_te/vrf_policy_driven_te_test.go:
```diff
@@ -0,0 +1,1946 @@
+		fluent.NextHopGroupEntry().WithNetworkInstance(deviations.DefaultNetworkInstance(dut)).
+			WithID(50).AddNextHop(50, 1),
+		fluent.IPv4Entry().WithNetworkInstance(niDecapTeVrf).
+			WithPrefix(prefWithMask).WithNextHopGroup(50),
```
Similarly please add WithNextHopGroupNetworkInstance(deviations.DefaultNetworkInstance(dut) here as well

cprabha replied:
same as above automated as per readme

arvbaska1 commented on feature/gribi/otg_tests/vrf_policy_driven_te/vrf_policy_driven_te_test.go:
```diff
@@ -0,0 +1,1946 @@
+		fluent.NextHopGroupEntry().WithNetworkInstance(deviations.DefaultNetworkInstance(dut)).
+			WithID(51).AddNextHop(51, 1),
+		fluent.IPv4Entry().WithNetworkInstance(niTeVrf111).
+			WithPrefix("0.0.0.0/0").WithNextHopGroup(51),
```
Similarly please add WithNextHopGroupNetworkInstance(deviations.DefaultNetworkInstance(dut) here as well

cprabha replied:
automated as per README

arvbaska1 commented on feature/gribi/otg_tests/vrf_policy_driven_te/vrf_policy_driven_te_test.go:
```diff
@@ -0,0 +1,1946 @@
+			WithIPinIP(ipv4OuterSrc111, gribiIPv4EntryVRF1112).
+			WithNextHopNetworkInstance(niTeVrf111),
+		fluent.NextHopGroupEntry().WithNetworkInstance(deviations.DefaultNetworkInstance(dut)).
+			WithID(11).AddNextHop(201, 1).AddNextHop(202, 3).WithBackupNHG(18),
```
Here for the prefix in Encap VRF are supposed to have below Backup NHG as per readme ?

IPv4Entry {138.0.11.0/24 (ENCAP_TE_VRF_A)} -> NHG#101 (DEFAULT VRF) -> {
  {NH#101, DEFAULT VRF, weight:1},
  {NH#102, DEFAULT VRF, weight:3},
  **backup_next_hop_group: 200 // in case specific vendor implementation or bugs pruned the NHs.**
}

NHG#200 (Default VRF) {
  {NH#200, DEFAULT VRF, weight:1}
}
NH#200 -> {
    network_instance: "DEFAULT"
}

arvbaska1 commented on feature/gribi/otg_tests/vrf_policy_driven_te/vrf_policy_driven_te_test.go:
```diff
@@ -0,0 +1,1946 @@
+		chk.IgnoreOperationID(),
+	)
+
+	// Install an 0/0 static route in ENCAP_VRF_A and ENCAP_VRF_B pointing to the DEFAULT VRF.
```
Should this be an actual static route instead of gRIBI route for the Encap VRF fallback ?
Although OC local routing does'nt specify a leaf NH VRF leaf in the model , there is a VRF fallback leaf in OC NI model.

cprabha replied:
Its as per README

* Install an 0/0 static route in ENCAP_VRF_A and ENCAP_VRF_B pointing to the DEFAULT VRF.

arvbaska1 replied:
Yes, but its not a Static route, its a gRIBI programmed route here.

arvbaska1 commented on feature/gribi/otg_tests/vrf_policy_driven_te/vrf_policy_driven_te_test.go:
```diff
@@ -0,0 +1,1946 @@
+	port7 := config.Ports().Add().SetName("port7")
+	port8 := config.Ports().Add().SetName("port8")
+
+	iDut1Dev := config.Devices().Add().SetName(atePort1.Name)
```
AddToOTG method can be used here

cprabha replied:
This one is working .

arvbaska1 commented on feature/gribi/otg_tests/vrf_policy_driven_te/vrf_policy_driven_te_test.go:
```diff
@@ -0,0 +1,1946 @@
+			// Send both 6in4 and 4in4 packets. Verify that the packets have their outer
+			// v4 header stripped and are forwarded according to the route in the DEFAULT
+			// VRF that matches the inner IP address.
+			portList := []string{"port8"}
```
This could be defined in global var maybe since its used in most tests

cprabha replied:
Please check we are using different port list depending on test.  Its not common across tests.

./vrf_policy_driven_te_test.go:494:     portList := []*ondatra.Port{p1, p2, p3, p4, p5, p6, p7, p8}
./vrf_policy_driven_te_test.go:1476:                    portList := []string{"port8"}
./vrf_policy_driven_te_test.go:1548:                    portList := []string{"port8"}
./vrf_policy_driven_te_test.go:1595:    portList := []string{"port8"}
./vrf_policy_driven_te_test.go:1644:            portList := []string{"port2"}
./vrf_policy_driven_te_test.go:1713:    portList := []string{"port8"}
./vrf_policy_driven_te_test.go:1789:            portList := []string{"port2"}

arvbaska1 commented on feature/gribi/otg_tests/vrf_policy_driven_te/vrf_policy_driven_te_test.go:
```diff
@@ -0,0 +1,1946 @@
+			0.1406, // 14.06 Port3
+			0.5625, // 56.25 Port4
+			0.25,   // 25 Port5
+			0,      // 0 Port6
```
This should be Port6 as well, not Port5 ?

cprabha replied:
Fixed.

arvbaska1 commented on feature/gribi/otg_tests/vrf_policy_driven_te/vrf_policy_driven_te_test.go:
```diff
@@ -0,0 +1,1946 @@
+
+					flow1 := createFlow(&flowArgs{flowName: flow4in4, srcPort: atePort1, dstPortList: dstPorts,
+						outHdrSrcIP: ipv4OuterSrc111, outHdrDstIP: ipv4OuterDst111, outHdrDscp: []uint32{dscpEncapA1},
+						InnHdrSrcIP: atePort1.IPv4, InnHdrDstIP: ipv4InnerDstNoEncap, isInnHdrV4: true})
```
Inner IP 138.0.11.12 will match with the 138.0.11.0/24 route in Encap VRF installed with base gRIBI programming ?
Need to use a different inner IP or maybe flush encap vrf gRIBI entries.

cprabha replied:
we are routing here based on pbf and it should take ni based on policy configured.

arvbaska1 replied:
Yes , so here its supposed match on source, protocol and DSCP, VRF_DECAP hit -> VRF_ENCAP_A miss -> DEFAULT ?
But in the ENCAP VRF_A there is already a 138.0.11.0/24 route, so there will be no route match miss right ?

arvbaska1 commented on feature/gribi/otg_tests/vrf_policy_driven_te/vrf_policy_driven_te_test.go:
```diff
@@ -0,0 +1,1946 @@
+	isisPort2V6.Addresses().Add().SetAddress(otgIsisPort8LoopV6).SetPrefix(uint32(128))
+
+	iDutBgp := iDut8Dev.Bgp().SetRouterId(otgIsisPort8LoopV4)
+	iDutBgp4Peer := iDutBgp.Ipv4Interfaces().Add().SetIpv4Name(iDut8LoopV4.Name()).Peers().Add().SetName(atePort8.Name + ".BGP4.peer")
```
is iBGP over loopback interface required ? same scenario can work with iBGP over connected interfaces

cprabha replied:
It is done as per README

arvbaska1 replied:
in the readme it just says BGP route resolved by ISIS in default VRF, so iBGP session can be over connected interface as well.

arvbaska1 commented on feature/gribi/otg_tests/vrf_policy_driven_te/README.md:
````diff
@@ -0,0 +1,837 @@
+1.  Add the following gRIBI entries:
+
+    ```
+    IPv4Entry {192.51.129.0/22 (DECAP_TE_VRF)} -> NHG#52 (DEFAULT VRF) -> {
````
the /22 prefix here should be 192.51.128.0/22 ?

cprabha replied:
Fixed.

arvbaska1 commented on feature/gribi/otg_tests/vrf_policy_driven_te/README.md:
```diff
@@ -0,0 +1,837 @@
+    * inner_dst: `ipv4_inner_encap_match`
+    * dscp: `dscp_encap_no_match`
+    * outter_src: `ipv4_outter_src_111`
+    * outter_dst: `192.51.100.64`
```
Traffic with IP DA = `192.51.100.64' will not match with any of prefixes installed in Decap vrf in this test ?

cprabha replied:
fixed.

cprabha commented on feature/gribi/otg_tests/vrf_policy_driven_te/vrf_policy_driven_te_test.go:
```diff
@@ -0,0 +1,1946 @@
+		Name:    "atePort3",
+		IPv4:    "192.0.2.10",
+		MAC:     "02:00:03:01:01:01",
+		IPv6:    "2001:db8::192:0:2:10",
```
updated.

cprabha commented on feature/gribi/otg_tests/vrf_policy_driven_te/vrf_policy_driven_te_test.go:
```diff
@@ -0,0 +1,1946 @@
+	dutPort4 = attrs.Attributes{
+		Desc:    "dutPort4",
+		IPv4:    "192.0.2.13",
+		IPv6:    "2001:db8::192:0:2:13",
```
updated

cprabha commented on feature/gribi/otg_tests/vrf_policy_driven_te/vrf_policy_driven_te_test.go:
```diff
@@ -0,0 +1,1946 @@
+	dutPort5 = attrs.Attributes{
+		Desc:    "dutPort5",
+		IPv4:    "192.0.2.17",
+		IPv6:    "2001:db8::192:0:2:17",
```
updated

cprabha commented on feature/gribi/otg_tests/vrf_policy_driven_te/vrf_policy_driven_te_test.go:
```diff
@@ -0,0 +1,1946 @@
+		Name:    "atePort8",
+		IPv4:    "192.0.2.30",
+		MAC:     "02:00:08:01:01:01",
+		IPv6:    "2001:db8::192:0:2:30",
```
updated

cprabha commented on feature/gribi/otg_tests/vrf_policy_driven_te/vrf_policy_driven_te_test.go:
```diff
@@ -0,0 +1,1946 @@
+	if deviations.InterfaceRefConfigUnsupported(dut) {
+		intf.InterfaceRef = nil
+	}
+	gnmi.Replace(t, dut, dutPolFwdPath.Config(), niP)
```
We are using single gnmi set req at the end , and I dont see any issues. It works very well.

cprabha commented on feature/gribi/otg_tests/vrf_policy_driven_te/vrf_policy_driven_te_test.go:
```diff
@@ -0,0 +1,1946 @@
+	}
+}
+
+func configGribiBaselineAFT(ctx context.Context, t *testing.T, dut *ondatra.DUTDevice, args *testArgs) {
```
I have used like other scripts are using.

cprabha commented on feature/gribi/otg_tests/vrf_policy_driven_te/vrf_policy_driven_te_test.go:
```diff
@@ -0,0 +1,1946 @@
+		fluent.NextHopGroupEntry().WithNetworkInstance(deviations.DefaultNetworkInstance(dut)).
+			WithID(7).AddNextHop(4, 1),
+		fluent.IPv4Entry().WithNetworkInstance(niTeVrf222).
+			WithPrefix(gribiIPv4EntryVRF2221+"/"+maskLen32).WithNextHopGroup(7),
```
Its not required, please refer https://github.com/openconfig/featureprofiles/blob/838865dd2897cede02bc834078d158f2d9c9fc93/feature/gribi/otg_tests/ipv4_entry_test/ipv4_entry_test.go#L180

arvbaska1 replied:
In above test the route is installed in Default NI itself, but e.g here https://github.com/openconfig/featureprofiles/blob/main/feature/gribi/otg_tests/base_hierarchical_nhg_update/base_hierarchical_nhg_update_test.go#L796 where the route is programmed in a non-default NI , then WithNextHopGroupNetworkInstance needs to be specified

cprabha commented on feature/gribi/otg_tests/vrf_policy_driven_te/vrf_policy_driven_te_test.go:
```diff
@@ -0,0 +1,1946 @@
+		fluent.NextHopGroupEntry().WithNetworkInstance(deviations.DefaultNetworkInstance(dut)).
+			WithID(9).AddNextHop(5, 2),
+		fluent.IPv4Entry().WithNetworkInstance(niTeVrf222).
+			WithPrefix(gribiIPv4EntryVRF2222+"/"+maskLen32).WithNextHopGroup(9),
```
please check readme , its automated as per readme.

cprabha commented on feature/gribi/otg_tests/vrf_policy_driven_te/vrf_policy_driven_te_test.go:
```diff
@@ -0,0 +1,1946 @@
+	}
+
+	// Programming AFT entries for prefixes in TE_VRF_222
+	args.client.Modify().AddEntry(t,
```
Its not there in README.md. please check README

arvbaska1 replied:
Its there in Readme here ? https://github.com/openconfig/featureprofiles/blob/main/feature/gribi/vrf_policy_driven_te/README.md?plain=1#L495 
https://github.com/openconfig/featureprofiles/blob/main/feature/gribi/vrf_policy_driven_te/README.md?plain=1#L531

sezhang2 commented:
https://github.com/openconfig/featureprofiles/pull/2539  has been approved

sezhang2 reviewed (dismissed):
It is an experimental code with many code commented out. 

We need review it after the implementation is completed and the test code is moved to offifical fiolder.

NishaSadhasivam-work reviewed (changes requested):
configureDUT configures port1-8 on the DUT.

But in Metadata.testproto TESTBED_DUT_ATE_2LINKS is used.

Please modify the same

NishaSadhasivam-work commented on feature/experimental/gribi/otg_tests/vrf_policy_driven_te/README.md:
```diff
@@ -0,0 +1,839 @@
+                            }
+                        }
+                        rule {
+                            sequence-id: 17
```
Can all the comments from https://github.com/openconfig/featureprofiles/pull/2217 be addressed in the readme file.

cprabha commented:
> configureDUT configures port1-8 on the DUT.
> 
> But in Metadata.testproto TESTBED_DUT_ATE_2LINKS is used.
> 
> Please modify the same

@NishaSadhasivam-work  I have submitted pull to update 8 port topology https://github.com/openconfig/featureprofiles/pull/2591 and update metadata for 8port.

NishaSadhasivam-work reviewed (approved):
LGTM!

arvbaska1 commented:
Hi @cprabha ,

Can you please address the comments for adding WithNextHopGroupNetworkInstance in gRIBI programming , where prefix is installed in non-default network-instance ?

cprabha commented:
> Hi @cprabha ,
> 
> Can you please address the comments for adding WithNextHopGroupNetworkInstance in gRIBI programming , where prefix is installed in non-default network-instance ?

@arvbaska1 , I will update it.

NishaSadhasivam-work reviewed (dismissed):
LGTM!

NishaSadhasivam-work reviewed (dismissed):
LGTM!

NishaSadhasivam-work reviewed (dismissed):
LGTM!

arvbaska1 commented:
> > Hi @cprabha ,
> > Can you please address the comments for adding WithNextHopGroupNetworkInstance in gRIBI programming , where prefix is installed in non-default network-instance ?
> 
> @arvbaska1 , I have updated , please check.

Thanks, can you also please add for route installed in decap VRF in funcconfigureGribiRoute & configureGribiMixedPrefEntries

Also similar to installing IPv4 0/0 default route in ENCAP_VRFs , an IPv6 ::/0 also needs to be added

cprabha commented:
> > > Hi @cprabha ,
> > > Can you please address the comments for adding WithNextHopGroupNetworkInstance in gRIBI programming , where prefix is installed in non-default network-instance ?
> > 
> > 
> > @arvbaska1 , I have updated , please check.
> 
> Thanks, can you also please add for route installed in decap VRF in func configureGribiRoute & configureGribiMixedPrefEntries
> 
> Also similar to installing IPv4 0/0 default route in ENCAP_VRFs , an IPv6 ::/0 also needs to be added

This is updated , please check.

arvbaska1 commented:
> > > > Hi @cprabha ,
> > > > Can you please address the comments for adding WithNextHopGroupNetworkInstance in gRIBI programming , where prefix is installed in non-default network-instance ?
> > > 
> > > 
> > > @arvbaska1 , I have updated , please check.
> > 
> > 
> > Thanks, can you also please add for route installed in decap VRF in func configureGribiRoute & configureGribiMixedPrefEntries
> > Also similar to installing IPv4 0/0 default route in ENCAP_VRFs , an IPv6 ::/0 also needs to be added
> 
> This is updated , please check.

Thanks. For IPv6 default route prefix should be ::/0 instead of 0::0/0 ?

cprabha commented:
> > > > > Hi @cprabha ,
> > > > > Can you please address the comments for adding WithNextHopGroupNetworkInstance in gRIBI programming , where prefix is installed in non-default network-instance ?
> > > > 
> > > > 
> > > > @arvbaska1 , I have updated , please check.
> > > 
> > > 
> > > Thanks, can you also please add for route installed in decap VRF in func configureGribiRoute & configureGribiMixedPrefEntries
> > > Also similar to installing IPv4 0/0 default route in ENCAP_VRFs , an IPv6 ::/0 also needs to be added
> > 
> > 
> > This is updated , please check.
> 
> Thanks. For IPv6 default route prefix should be ::/0 instead of 0::0/0 ?

updated.

arvbaska1 reviewed (approved):
Looks good for us.
We will check for iBGP over loopback vs physical interface, & open a PR if needed.

NishaSadhasivam-work reviewed (approved):
LGTM!
